### ISSUE TITLE
AcquireTokenSilently without Specifying Scopes

### ISSUE THREAD

**Issue by javate** [AzureAD/microsoft-authentication-library-for-java#375]
I need to refresh the access token but I don't want to specify `scopes` in `SilentParameters`. I want to refresh it for all possible scopes which are founded in `TokenCache` object also.

So, can I reach scopes in the `TokenCache`? and then I can pass s `SilentParameters`. Or is there another way for this?

For example like below, It was passed `"User.Read"` however I don't want to specify any scope. I only want refresh my current access token(which is also founded in `TokenCache`) silently. How can I do that? 
```
app.tokenCache().deserialize(tokenCache.toString());
SilentParameters parameters = SilentParameters.builder( Collections.singleton("User.Read"),result.account()).build();
CompletableFuture<IAuthenticationResult> future = app.acquireTokenSilently(parameters);
IAuthenticationResult updatedResult = future.get();
```

**Comment by Avery-Dunn** [comment]
Hello @javate : I don't believe there's an easy way to get the scopes for a given access token. The scopes are saved in the `target` field of access tokens in the token cache, but you'd need to parse the cache yourself to get that info.

Could you describe your use case a bit here? [Scopes](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent#scopes-and-permissions) are basically just the way to request permissions for tokens to have, and applications should know what permissions it needs to access resources for users. If your application has some access tokens in the cache, then presumably your application specified the scopes those tokens have.

**Comment by javate** [comment]
Hi @Avery-Dunn 
In my use case, I will expose an API like "getGainedScopes". This API gives me the gained scopes of the user until that time. So according to this info, I can understand which scopes are missing for my purpose than I can create an Authorize URL(pop up url) by using the missing scopes for satisfying all required scopes.

Also, I was try to

**Comment by Avery-Dunn** [comment]
Just to be clear, you're trying to find what scopes a certain user's access tokens already have, so you can then authorize a user for only the scopes that they don't already have an access token for? If so, I don't believe that's a recommended way of doing token requests (or at least I haven't heard of someone doing that before), so there isn't really a good way to do that in our library. If you need certain scopes to access some resource, you should just request a token for the specific set of scopes you need to access that resource.

As for sending blank scopes in a silent call, you're seeing those scopes in the AuthenticationResult object the silent call returns, right? The relevant behavior is in the `TokenCache` class, mainly in the `getAccessTokenCacheEntity` method [here](https://github.com/AzureAD/microsoft-authentication-library-for-java/blob/1c89680337d20eb90b721af0f77fa5422fb38790/src/main/java/com/microsoft/aad/msal4j/TokenCache.java#L470). Basically, when trying to find an access token from the cache we filter the access tokens based on a number of things, including scopes. 

However, it only returns a cached access token if it has all of the scopes you send in the silent call, and if it doesn't match then the cached access token isn't returned. Instead, it attempts to 'refresh' the cached access token using the scopes you sent in the silent request, so the scopes you're seeing in the result are actually the scopes in a new access token, not the ones in a cached token.

**Comment by javate** [comment]
Thanks, I understand. 
You said "it attempts to 'refresh' the cached access token using the scopes you sent in the silent request" then what would happen if I send blank scopes?

**Comment by Avery-Dunn** [comment]
I wasn't sure about the behavior of sending blank scopes, and also couldn't find any documentation covering it. I did some testing and it seems like you're describing it correctly: I made a request with some scopes, then a request with blank scopes and got that first set returned, tried another request with a new scope, and then blank scopes again and got the original set plus that new scope.

I couldn't find this documented anywhere, so one disadvantage to this approach is that it's behavior from the token service and not MSAL Java, so our library can't really guarantee that this behavior won't change. Additionally, since this seems to return access tokens with all of the user's permissions you'll only be able to use the token cache/silent calls when you make a request with all of those scopes.
